File: backend/app/core/upgrades/stat_modifiers/explosion_damage_percent_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class ExplosionDamagePercent2(UpgradeImplementation):
    id = 'explosion_damage_percent_2'
    name = 'Explosive Force'
    description = '+7% explosion damage'
    rarity = 'legendary'
    type = 'stat_modifier'
    cost = 40

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.07
        appliers.modifiers.add_multiplier('bullet', 'explosionDamage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/explosion_damage_percent_2.ts
import { Upgrade, type UpgradeDef, type ExplosionSpec } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const ExplosionDamagePercent2Def: UpgradeDef = {
  id: "explosion_damage_percent_2",
  name: "Explosive Force",
  description: "+7% explosion damage",
  rarity: RarityID.Legendary,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 40,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.ExplosionDamage,
  value: 0.07,
  isMultiplier: true,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 3,
  dependentOn: ["explosive_bullets", "explosion_on_kill"],
  dependencyCount: 1,
}

export class ExplosionDamagePercent2 extends Upgrade {
  // Stat fields in the def are UI/backend metadata only — the behavior is the
  // hook below, so the default modifier-channel application is skipped.
  onApply(): void {}

  modifyExplosion(explosion: ExplosionSpec): void {
    explosion.damage *= 1 + this.def.value!
  }
}
